reco_utils/dataset/pandas_df_utils.py

```diff
@@ -74,3 +75,75 @@ def filter_by(df, filter_by_df, filter_by_cols):
             filter_by_df.set_index(filter_by_cols).index
         )
     ]
+
+
+def df_to_libffm(df, col_rating=DEFAULT_RATING_COL, filepath=None):
+    """Converts an input Dataframe (df) to an text file in libffm format.
+
+    Note:
+        The input dataframe is expected to represent the feature data in the following schema
+        |field-1|field-2|...|field-n|rating|
```

Comment:
I think this is too restrictive. You require that all fields have k features. Most of the time the number of features differs across fields.

Reply:
Yea this is a good point - I am thinking whether we want to put it into some auxiliary functions that perform the data preparation if the original data is not in the expected format. What do you think?

Reply:
They have more examples in the slides with varying sizes. I guess I do not understand the purpose of this function. You don't really need fields in this case, just features. To have both fields and features in the representation, you would need the user to specify the mapping of features to fields, or to identify the fields with the categorical features (how you obtain these from a pandas DF is yet another discussion).

Reply:
Yea this is why I think the input of the function is standardized so that the field and feature information can be derived from the Dataframe directly. I think the purpose of the function is to help map the field and feature from the information in the Dataframe to another which is represented in `field:feature:1` or `field:field:value` format.

Reply:
See my comments below.

Reply:
xDeepFM uses the FFM format as data input: <label> <field_id>:<feature_id>:<feature_value> I guess it is the same for other deeprec algorithms.

Reply:
> xDeepFM uses the FFM format as data input: <field_id>:<feature_id>:<feature_value> I guess it is the same for other deeprec algorithms.

If **feature_value** here in the <field_id>:<feature_id>:<feature_value> I is **1** or **the actual feature value** (if feature here is the one used in the original paper), yea, this is what the function does.